sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala

```diff
@@ -218,9 +221,44 @@ class AnalysisSuite extends AnalysisTest {
 
     // CreateStruct is a special case that we should not trim Alias for it.
     plan = testRelation.select(CreateStruct(Seq(a, (a + 1).as("a+1"))).as("col"))
-    checkAnalysis(plan, plan)
+    expected = testRelation.select(CreateNamedStruct(Seq(
+      Literal(a.name), a,
+      Literal("a+1"),(a + 1))).as("col"))
+    checkAnalysis(plan, expected)
     plan = testRelation.select(CreateStructUnsafe(Seq(a, (a + 1).as("a+1"))).as("col"))
-    checkAnalysis(plan, plan)
+    expected = testRelation.select(CreateNamedStructUnsafe(Seq(
+      Literal(a.name), a,
+      Literal("a+1"),(a + 1))).as("col"))
+    checkAnalysis(plan, expected)
+  }
+
+  test("Analysis may leave unnecassary aliases") {
+    val att1 = testRelation.output.head
+    var plan = testRelation.select(
+      CreateStructUnsafe(Seq(att1, ((att1 as "aa") + 1).as("a_plus_1"))).as("col"),
+      att1
+    )
+    val prevPlan = getAnalyzer(true).execute(plan)
+    plan = prevPlan.select(CreateArray(Seq(
+      CreateStructUnsafe(Seq(att1, (att1 + 1).as("a_plus_1"))),
+      /** alias should be eliminated by [[CleanupAliases]] */
+      "col".attr as "col2"
+    )) as "arr")
+    plan = getAnalyzer(true).execute(plan)
+
+    plan should be (a[Project])
+    val Project( projectExpressions, _) = plan
+    projectExpressions should have (size(1))
+    val Seq(expr1) = projectExpressions
+    expr1 should be (a[Alias])
+    val Alias(expr2, arrAlias) = expr1
+    arrAlias shouldBe "arr"
+    expr2 should be (a[CreateArray])
+    val CreateArray(arrElements) = expr2
+    arrElements should have (size(2))
+    val Seq(el1, el2) = arrElements
+    el1 should not be a[Alias]
```

Comment:
when using the 'should' dsl, parenthesis are better left aside, however there's a limit to what you can do with a DSL under scala. in previous cases the 'should' sentences have an even number of words, so scala requires the last one to be within parenthesis. in this line we have an odd number of words sp we can get away with it.